### ISSUE TITLE
Epic: adversarial-patch-finding-on-a-small-vision-language-model

### ISSUE THREAD

**Issue by rize0528** [rize0528/ccpm_handson#1]
# Epic: Adversarial Patch Finding on a Small Vision-Language Model

## Overview

A standalone Python system that processes motion JPEG streams in real-time to detect adversarial patches targeting small vision-language models. The system combines statistical anomaly detection with VLM performance monitoring to provide binary attack classification per frame, designed for research demonstrations and security assessments.

## Architecture Decisions

**Core Technology Stack:**
- **Python 3.9+** for development simplicity and ML library compatibility
- **PyTorch** as primary deep learning framework (better adversarial tooling than TensorFlow)
- **OpenCV** for motion JPEG processing and computer vision operations
- **CLIP (OpenAI)** as the target small VLM (pre-trained, well-documented, lightweight)
- **NumPy/SciPy** for statistical analysis and feature processing

**Key Design Patterns:**
- **Pipeline Architecture**: Modular components for stream processing, detection, and output
- **Strategy Pattern**: Pluggable detection methods (statistical, ensemble-based, gradient-based)
- **Observer Pattern**: Real-time monitoring of VLM performance degradation
- **Factory Pattern**: Different patch generators for various attack types

**Detection Approach:**
- Primary: Statistical anomaly detection on CLIP image embeddings
- Secondary: Monitor VLM confidence score degradation
- Tertiary: Frequency domain analysis for physical patch artifacts

### Core Components

**Stream Processing Engine**
- Motion JPEG decoder using OpenCV VideoCapture
- Frame buffer management with configurable size
- Asynchronous processing pipeline to meet <200ms latency
- Input validation and error handling for malformed streams

**VLM Integration Layer**
- CLIP model wrapper with consistent preprocessing
- Batch processing optimization for efficiency
- Feature extraction pipeline for both image and text embeddings
- Performance baseline establishment during initialization

**Detection Engine**
- Multi-method detection using weighted voting
- Feature-based anomaly detection (isolation forest, one-class SVM)
- Gradient magnitude analysis for adversarial perturbations
- Ensemble approach combining multiple detection signals

**Patch Generator Module**
- PGD (Projected Gradient Descent) implementation
- C&W (Carlini & Wagner) attack variants
- Physical patch optimization with geometric transformations
- Targeted attacks against specific CLIP capabilities

### Backend Services

**Core API Endpoints:**
- `POST /stream/start` - Initialize stream processing
- `GET /stream/status` - Current processing status
- `POST /stream/stop` - Stop processing and cleanup
- `GET /detection/results` - Retrieve detection results
- `POST /patch/generate` - Generate adversarial patches for testing

**Data Models:**
```python
@dataclass
class DetectionResult:
    frame_id: int
    timestamp: float
    is_attack: bool
    confidence_score: float
    vlm_degradation: float

@dataclass
class ProcessingConfig:
    stream_source: str
    detection_threshold: float
    buffer_size: int
    target_fps: int
```

### Infrastructure

**Deployment Considerations:**
- Single-process application with multi-threading for real-time processing
- GPU support optional but recommended for CLIP inference
- Memory-efficient frame buffering to handle continuous streams
- Graceful degradation when GPU unavailable

**Monitoring and Observability:**
- Processing latency metrics per frame
- Detection accuracy tracking against known test cases
- VLM inference time monitoring
- Memory and CPU utilization tracking

## Implementation Strategy

**Development Phases:**

**Phase 1: Core Detection Pipeline (Weeks 1-2)**
- Implement motion JPEG stream processing
- Integrate CLIP model for baseline VLM behavior
- Build statistical anomaly detection on image embeddings
- Create binary output formatting

**Phase 2: Patch Generation & Testing (Weeks 3-4)**
- Implement PGD adversarial patch generation
- Add C&W attack variants
- Create test dataset with generated patches
- Validate detection accuracy against generated attacks

**Phase 3: Physical Optimization & Robustness (Weeks 5-6)**
- Optimize patches for physical deployment (lighting, angles)
- Enhance detection for physical patch characteristics
- Performance optimization to meet latency requirements
- End-to-end testing with camera feeds

**Risk Mitigation:**
- Start with pre-trained CLIP to avoid model training complexity
- Use well-established adversarial attack libraries (Advertorch, Foolbox)
- Implement fallback detection methods if primary approach fails
- Comprehensive unit testing for each pipeline component

**Testing Approach:**
- Unit tests for each module with mock data
- Integration tests with known adversarial examples
- Performance benchmarking against latency requirements
- End-to-end testing with real motion JPEG streams

## Task Breakdown Preview

High-level task categories that will be created:
- [ ] **Environment Setup**: Python environment, dependencies, CLIP model integration
- [ ] **Stream Processing**: Motion JPEG decoder, frame buffering, async pipeline
- [ ] **VLM Integration**: CLIP wrapper, feature extraction, baseline establishment
- [ ] **Detection Engine**: Statistical anomaly detection, multi-method ensemble
- [ ] **Patch Generator**: PGD/C&W implementation, physical optimization
- [ ] **API Layer**: REST endpoints, data models, error handling
- [ ] **Testing Framework**: Unit tests, integration tests, performance benchmarks
- [ ] **Physical Validation**: Camera integration, real-world testing
- [ ] **Performance Optimization**: Latency optimization, memory efficiency
- [ ] **Documentation**: API docs, usage guide, demo setup instructions

## Dependencies

**External Service Dependencies:**
- OpenAI CLIP model (pre-trained weights from Hugging Face)
- COCO dataset for clean image baselines
- ImageNet subset for diverse testing scenarios

**Internal Team Dependencies:**
- Access to camera hardware for physical patch testing
- GPU resources for model inference and patch generation
- Test environment setup for continuous integration

**Prerequisite Work:**
- Python development environment with GPU support
- Dataset download and preprocessing pipeline
- Basic familiarity with adversarial ML concepts

## Success Criteria (Technical)

**Performance Benchmarks:**
- Frame processing latency: <200ms per frame (target: <100ms)
- Detection accuracy: >85% on generated adversarial patches
- False positive rate: <10% on clean COCO/ImageNet images
- System stability: Run continuously for 30+ minutes without crashes

**Quality Gates:**
- 90%+ unit test coverage for core modules
- Integration tests passing for all major workflows
- Performance benchmarks meeting or exceeding targets
- Code review approval with security considerations

**Acceptance Criteria:**
- Binary classification output for each processed frame
- Successful processing of various motion JPEG formats
- Demonstration-ready system with simple CLI interface
- Clear documentation for setup and usage

## Estimated Effort

**Overall Timeline:** 6 weeks (single developer, part-time equivalent)

**Resource Requirements:**
- Development machine with GPU support (GTX 1060 or better)
- 20-30 hours per week development time
- Access to camera for physical testing (final 2 weeks)

**Critical Path Items:**
1. CLIP integration and baseline establishment (Week 1)
2. Statistical detection method implementation (Week 2)
3. Adversarial patch generation pipeline (Week 3)
4. Physical patch optimization and testing (Weeks 5-6)

**Risk Factors:**
- CLIP model performance on target hardware
- Adversarial patch generation convergence
- Physical patch effectiveness in real environments
- Meeting strict latency requirements with current hardware

## Stats

Total tasks: 9
Parallel tasks: 5 (can be worked on simultaneously)
Sequential tasks: 4 (have dependencies)
Estimated total effort: 140 hours
